cgi-bin/DW/API/Key.pm

```diff
@@ -0,0 +1,191 @@
+# or undef if the user has no API keys yet.
+sub get_keys_for_user {
+    my ($self, $u) = @_;
+    my $user = LJ::want_user( $u )
+        or die "need a user!\n";
+    my $memkey = [ $user->id, "api_keys_list:" . $user->id ];
+    my $keydata = LJ::MemCache::get( $memkey );
+    my @keylist;
+
+    if ( defined $keydata ) {
+        for my $keyhash (@$keydata) {
+            push @keylist, ($self->get_key($keyhash));
+        }
+
+    } else {
+        my $dbr = LJ::get_db_reader() or die "Failed to get database";
+        my $keys = $dbr->selectall_hashref( 
+                q{SELECT keyid, hash FROM api_key WHERE userid = ? AND state = 'A'},
+                'keyid', undef, $user->{userid} 
+            );
+        carp $dbr->errstr if $dbr->err;
+        return undef unless $keys;
+        my @hashlist;
+
+        for my $key (sort (keys %$keys)) {
+            my $new = $self->_create($user, $keys->{$key}->{keyid}, $keys->{$key}->{hash});
+            push @hashlist, ($keys->{$key}->{hash});
+            push @keylist, ($new);
+            my $cachekey = [ $new->{keyhash}, "api_key:" . $new->{keyhash} ];
+            LJ::MemCache::set( $cachekey, { userid => $user->id, keyid => $new->{keyid}, hash => $new->{keyhash}} );
+        }
+        LJ::MemCache::set( $memkey, \@hashlist );
+    }
+
+    return \@keylist;
+}
+
+
+
+# Usage: create ( user, key ) 
+# Creates and returns a new key object given a user and key hash.
+# Don't call this directly, as it neither verifies nor saves keys.
+# new() or lookup() is probably what you want instead. 
+sub _create {
+    my ($class, $user, $keyid, $keyhash) = @_;
+
+    my %key = (
+        user => $user,
+        keyid => $keyid,
+        keyhash => $keyhash
+        );
+
+    bless \%key, $class;
+    return \%key;
+}
+
+# Usage: $key->can_read( resource ) 
+# Checks if a key has been given read permissions
+# for a given resource by the user it belongs to.
+
+sub can_read {
+    my ($self, $resource) = @_;
+
+    #TODO: Once key scoping is implemented, actually check this
+    return 1;
+}
+
+# Usage: $key->can_write( resource ) 
+# Checks if a key has been given write permissions
+# for a given resource by the user it belongs to.
+
+sub can_write {
+    my ($self, $resource) = @_;
+
+    #TODO: Once key scoping is implemented, actually check this
+    return 1;
+}
+
+# Usage: $key->delete ($user) 
+# Marks a key as deleted in the DB. A user is required to guarantee
+# that the key is being deleted by someone with the permission to do so.
+sub delete {
+    my ($self, $u) = @_;
+    my $user = LJ::want_user( $u )
+        or die "need a user!\n";
+    my $memkey = [ $self->{keyhash}, "api_key:" . $self->{keyhash} ];
+
+    $self->valid_for_user($user) or die "key doesn't belong to user";
+
+    my $dbw = LJ::get_db_writer() or die "Failed to get database";
+    $dbw->do(
+        q{UPDATE api_key SET state = 'D' WHERE hash = ?},
```

Comment:
Today there's only two states, but I'd traditionally like to have the WHERE clause contain the from state we want to be true just to avoid race conditions. So, like `WHERE state = 'A' AND hash = ?`.

Reply:
This didn't get done.